### ISSUE TITLE
Gen.listOfN refuses to generate stuff

### ISSUE THREAD

**Issue by ostronom** [typelevel/scalacheck#256]
``` .scala
scala> import org.scalacheck._
import org.scalacheck._

scala> :paste
// Entering paste mode (ctrl-D to finish)

object Generators {
  val unicodeChar: Gen[Char] = Gen.choose(Char.MinValue, Char.MaxValue).filter(c ⇒ Character.isDefined(c))

  def unicodeStr(a: Int, b: Int): Gen[String] =
    for {
      n ← Gen.chooseNum(a, b).map { rnd ⇒
        println(s"generated $rnd")
        rnd
      }
      s ← Gen.listOfN(n, unicodeChar).map { s ⇒
        val str = s.mkString
        println(s"generated $str for $n")
        str
      }.retryUntil(_.length == n)
    } yield s
}


// Exiting paste mode, now interpreting.

defined object Generators

scala> Generators.unicodeStr(1,5).sample.get
generated 3
generated ｵ輜 for 3
res0: String = ｵ輜

scala> Generators.unicodeStr(256,1024).sample.get
generated 286
java.util.NoSuchElementException: None.get
  at scala.None$.get(Option.scala:347)
  at scala.None$.get(Option.scala:345)
  ... 42 elided
```

What is wrong with my code? It works well for all n up to 255. As far as i know, Int can be greater than 255..

**Comment by non** [comment]
@ostronom The issue here is your use of `.filter`. Following the chain of calls, we can see:

 - `Gen.listOfN`
 - `Gen.buildableOfN`
 - `Gen.sequence`

The documentation for sequence says:

> If any of the given generators fails, the resulting generator will also fail.

In my tests, about 3.5% of the characters generated in `unicodeChar` will fail to be valid. This means that for each additional character you try to add to a list, you have an additional 3.5% chance of failure. Alternately phrased, each character has a 96.5% chance of success, so for length N the chance of success is (0.965 ^ N). That means that there is a 2.8% chance of succeeding for N=100.

My guess is that even with your `retryUntil` call, that ScalaCheck eventually bails out if it fails too many times. I can imagine three solutions here:

(1) Figure out a way to avoid `.filter`. For example, you could map invalid characters to known valid characters (e.g. `a`).

(2) Use a different method to construct your string than `listOfN` that is less likely to fail.

(3) Write a PR to make `Gen.listOfN` less likely to fail in these cases.

Of all of these I think (1) is the easiest to do and the likely to succeed. Hope this helps!

**Comment by non** [comment]
@ostronom Does this make sense to you? It seems like we can close this issue, what do you think?

**Comment by Stefanqn** [comment]
to be honest, I think the current implementation / behavior is not intuitive.
this worked for me:
```
  val elemListGen = {
    for (
      max <- choose(1, 500);
      i <- (1 to max).toList;
      e <- elemGen.sample
    ) yield e 
  }
```
where elemGen fails often bc it uses suchThat
```
  val elemGen = {
    for (
      n <- containerOfN[Set, Int](LotteryTicket.numberSize, numberGen) suchThat { _.size == LotteryTicket.numberSize };
      s <- containerOfN[Set, Int](LotteryTicket.superNumberSize, superNumberGen) suchThat { _.size == LotteryTicket.superNumberSize }
    ) yield LotteryTicket(n, s)
  }
```

**Comment by eyalfa** [comment]
@non,
I think I've managed to come up with a better Impl for listOf/listOfN/nonEmptyListOf, I've described it in the users list: https://groups.google.com/forum/m/#!topic/scalacheck/7MIkeT4XoCk.

Do you think a PR introducing this approach is likely to be accepted?

**Comment by ashawley** [comment]
I believe the test failures for discarded values for buildableOfN and containerOfN were fixed with #714 and released in version 1.15.2.

Running a similar test to the last example generator will pass now:

```scala
  val pairGen = {
    for (
      n <- Gen.containerOfN[Set, Int](5, Gen.choose(1,47)) suchThat { _.size == 5 };
      s <- Gen.containerOfN[Set, Int](1, Gen.choose(1,27)) suchThat { _.size == 1 }
    ) yield (n, s)
  }

  property("pairGen") = Prop.forAll(pairGen) { case (m, n) =>
    m.size == 5 && n.size == 1
  }

```

If this issue persists feel free to reopen.  Thanks for the bug report.
